client/circuitbreaker/circuit_breaker.go

```diff
@@ -0,0 +1,280 @@
+	}
+
+	defer func() {
+		e := recover()
+		if e != nil {
+			cb.onResult(state, Yes)
+			panic(e)
+		}
+	}()
+
+	result, overloaded, err := call()
+	cb.onResult(state, overloaded)
+	return result, err
+}
+
+func (cb *CircuitBreaker[T]) onRequest() (*State[T], error) {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+
+	state, err := cb.state.onRequest(cb)
+	cb.state = state
+	return state, err
+}
+
+func (cb *CircuitBreaker[T]) onResult(state *State[T], open Overloading) {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+
+	if cb.state == state {
+		state.onResult(open)
+	} // else the state moved forward so we don't need to update the counts
+}
+
+// State represents the state of CircuitBreaker.
+type State[T any] struct {
+	stateType StateType
+	cb        *CircuitBreaker[T]
+	end       time.Time
+
+	pendingCount uint32
+	successCount uint32
+	failureCount uint32
+}
+
+// newState creates a new State with the given configuration and reset all success/failure counters.
+func (cb *CircuitBreaker[T]) newState(now time.Time, stateType StateType) *State[T] {
+	var end time.Time
+	var pendingCount uint32
+	switch stateType {
+	case StateClosed:
+		end = now.Add(cb.config.ErrorRateWindow)
+	case StateOpen:
+		end = now.Add(cb.config.CoolDownInterval)
+	case StateHalfOpen:
+		// we transition to HalfOpen state on the first request after the cooldown period,
+		// so we start with 1 pending request
+		pendingCount = 1
+	default:
+		panic("unknown state")
+	}
+	return &State[T]{
+		cb:           cb,
+		stateType:    stateType,
+		pendingCount: pendingCount,
+		end:          end,
+	}
+}
+
+// onRequest transitions the state to the next state based on the current state and the previous requests results
+// All state transitions happens at the request evaluation time only
+// The implementation represents a state machine effectively
+func (s *State[T]) onRequest(cb *CircuitBreaker[T]) (*State[T], error) {
+	var now = time.Now()
+	switch s.stateType {
+	case StateClosed:
+		if s.end.Before(now) {
+			// ErrorRateWindow is over, let's evaluate the error rate
+			total := s.failureCount + s.successCount
+			observedErrorRatePct := s.failureCount * 100 / total
+			if s.cb.config.ErrorRateThresholdPct > 0 && total >= uint32(s.cb.config.ErrorRateWindow.Seconds())*s.cb.config.MinQPSForOpen && observedErrorRatePct >= s.cb.config.ErrorRateThresholdPct {
+				// the error threshold is breached, let's move to open state and start failing all requests
+				log.Error("Circuit breaker tripped. Starting to fail all requests",
+					zap.String("name", cb.name),
+					zap.Uint32("observedErrorRatePct", observedErrorRatePct),
+					zap.String("config", fmt.Sprintf("%+v", cb.config)))
+				cb.fastFailCounter.Inc()
+				return cb.newState(now, StateOpen), errs.ErrCircuitBreakerOpen
+			}
+			// the error threshold is not breached or there were not enough requests to evaluate it,
+			// continue in the closed state and allow all requests
+			return cb.newState(now, StateClosed), nil
```

Comment:
Do we need to return a new state🤔? It means if the fact error rate is not met `ErrorRateThresholdPct` during one `ErrorRateWindow`, we will move the window into next window. Isn't the range of movement too large? Can the window be moved in one second?

Reply:
> The time consists of windows back to back. Windows are static, they are not moving. When one window is over we evaluate the error rate, then we move to the next state with a new window.

OK, my previous idea was a movable window, but I think the current implementation can also meet the needs. LGTM